File: api/app/core/blog.py
from model.blog import Blog, Tag
from store.client import BlogClient
from store.impl.blog import PostgresBlogClient


class BlogCore:
    _blog_client: BlogClient

    def __init__(self, blog_client: BlogClient) -> None:
        self._blog_client = blog_client

    def get_blog(self, blog_id: int) -> Blog | None:
        try:
            return self._blog_client.get_blog(blog_id)
        except Exception as e:
            raise Exception(f'Failed to get blog: {e}')

    def get_all_blogs(self) -> list[Blog]:
        try:
            return self._blog_client.get_all_blogs()
        except Exception as e:
            raise Exception(f'Failed to get all blogs: {e}')

    def get_tag(self, tag_id: int) -> Tag | None:
        try:
            return self._blog_client.get_tag(tag_id)
        except Exception as e:
            raise Exception(f'Failed to get tag: {e}')

    def get_all_tags(self) -> list[Tag]:
        try:
            return self._blog_client.get_all_tags()
        except Exception as e:
            raise Exception(f'Failed to get all tags: {e}')

    def search_blogs_by_tag(self, tag_id: int) -> tuple[Tag, list[Blog]] | None:
        try:
            tag = self.get_tag(tag_id)
            if tag:
                blogs = self._blog_client.search_blogs_by_tag(tag_id)
                return (tag, blogs)
            else:
                return None
        except Exception as e:
            raise Exception(f'Failed to search blogs by tag: {e}')

    def search_blogs_by_year(self, year: int) -> list[Blog]:
        try:
            return self._blog_client.search_blogs_by_year(year)
        except Exception as e:
            raise Exception(f'Failed to search blogs by year: {e}')


blog_core = BlogCore(PostgresBlogClient())

File: frontend/core/fetch/blog.ts
import { parse } from 'date-fns'
import type { Blog, GetAllBlogs, GetAllTags, GetBlogById, GetTagById, Tag } from '@/core/model/blog'

type TagResponse = {
  tag_id: number,
  name: string,
  created_at: string,
  updated_at: string | null,
}

type BlogResponse = {
  blog_id: number,
  title: string,
  content: string,
  created_at: string,
  updated_at: string | null,
  tags: TagResponse[],
}

export const parseTimeIgnoreSubSecond = (time: string): Date => {
  return parse(time.substring(0, 19), "yyyy-MM-dd'T'HH:mm:ss", new Date())
}

export const parseNullableTime = (time: string | null): Date | null => {
  return time ? parseTimeIgnoreSubSecond(time) : null
}

export const transformTagResponse = (tag: TagResponse): Tag => ({
  tag_id: tag.tag_id,
  name: tag.name,
})

export const transformBlogResponse = (blog: BlogResponse): Blog => ({
  blog_id: blog.blog_id,
  title: blog.title,
  content: blog.content,
  created_at: parseTimeIgnoreSubSecond(blog.created_at),
  updated_at: parseNullableTime(blog.updated_at),
  tags: blog.tags.map(transformTagResponse),
})

export const getAllBlogs: GetAllBlogs = async () => {
  const res = await fetch(`${process.env.API_HOST}/v1/blog/all`)
  const json = await res.json()
  return json.blogs.map((blog: BlogResponse) => transformBlogResponse(blog))
}

export const getBlogById: GetBlogById = async (blog_id) => {
  const res = await fetch(`${process.env.API_HOST}/v1/blog/entry/${blog_id}`)
  const json = await res.json()
  return transformBlogResponse(json.blog)
}

export const getAllTags: GetAllTags = async () => {
  const res = await fetch(`${process.env.API_HOST}/v1/tag/all`)
  const json = await res.json()
  return json.tags.map((tag: TagResponse) => transformTagResponse(tag))
}

export const getTagById: GetTagById = async (tag_id) => {
  const res = await fetch(`${process.env.API_HOST}/v1/tag/entry/${tag_id}`)
  const json = await res.json()
  return transformTagResponse(json.tag)
}
